### PR TITLE
Add Hand-Eye calibration methods

### PR DESCRIPTION
**Pull request by catree**
### This pullrequest changes

Add Hand-Eye calibration from Tsai method. This allows estimating the rigid transformation between a camera mounted on a robot gripper:

hand-eye_figure

See also:
 - [Hand-Eye Calibration ](http://campar.in.tum.de/Chair/HandEyeCalibration)
 - [A new technique for fully autonomous and efficient 3D robotics hand/eye calibration,  R.Y. Tsai ; R.K. Lenz](https://pdfs.semanticscholar.org/19b3/89a797a55c8b63dca8b6d1889df4cff8bfaa.pdf)

The current interface is:
```
void calibrateHandEyeTsai( InputArrayOfArrays R_gripper2base, InputArrayOfArrays t_gripper2base,
                                            InputArrayOfArrays R_target2cam, InputArrayOfArrays t_target2cam,
                                            OutputArray R_cam2gripper, OutputArray t_cam2gripper );
```

 - currently there is only one method (Tsai). Should we keep this or use a more generic one `calibrateHandEye` with a method parameter to choose different calibration methods? This can be a more general question also (e.g. `calibrateCamera`, `calibrateCameraRO`, `fitEllipse`, `fitEllipseAMS`, `fitEllipseDirect`, ...).
 - inputs are rotation matrices and translation vectors. It could have been rotation and translation vectors
 - naming is consistent with the other functions, alternative is `HandEyeCalibration()`
 - code follows the paper notations
 - C++ code from this [Matlab code](http://lazax.com/www.cs.columbia.edu/~laza/html/Stewart/matlab/handEye.m)

### CODE REVIEW

File: modules/calib3d/src/calibration.cpp [hunk 1]
```diff
@@ -3855,6 +3855,91 @@ static void adjust3rdMatrix(InputArrayOfArrays _imgpt1_0,
...
+{
+    CV_Assert( T.rows == 4 && T.cols == 4 );
+
+    Mat R = T(Rect(0, 0, 3, 3));
+    Mat t = T(Rect(3, 0, 1, 3));
+    Mat Rt = R.t();
+    Mat tinv = -Rt * t;
+    Mat Tinv = Mat::eye(4, 4, T.type());
+    Rt.copyTo(Tinv(Rect(0, 0, 3, 3)));
+    tinv.copyTo(Tinv(Rect(3, 0, 1, 3)));
```
Comment (alalek) [inline modules/calib3d/src/calibration.cpp]: Consider using of `Matx44d` for well known mat sizes (including input/output parameters).

**Comment by catree** [comment]
I have added some well-known hand-eye calibration methods:
  - solve rotation then translation:
    - R. Tsai, R. Lenz A New Technique for Fully Autonomous and Efficient 3D Robotics Hand/EyeCalibration
    - F. Park, B. Martin Robot Sensor Calibration: Solving AX = XB on the Euclidean Group
    - R. Horaud, F. Dornaika Hand-Eye Calibration
  - solve simultaneously rotation and translation:
    - N. Andreff, R. Horaud, B. Espiau On-line Hand-Eye Calibration
    - K. Daniilidis Hand-Eye Calibration Using Dual Quaternions

I have also moved the code to a new file `calibration_handeye.cpp`.

Many thanks to Mili Shah for providing a clean [Matlab implementation](http://math.loyola.edu/~mili/Calibration/) for these algorithms.
> M. Shah, R. D. Eastman, T. Hong, An Overview of Robot-Sensor Calibration Methods for Evaluation of Perception Systems, Performance Metrics for Intelligent Systems, (2012). 

---

For now, I did not used `Matx44d` and explicit matrix shape since the conversion is not direct and requires more time and extra work.
Missing features or I did not found a way to do it:
  - matrix slicing
  - `Matx::copyTo()` to copy for example a `Matx44d` to a bigger matrix, something like:
```
Matx31d diff = Pcij - Pgij;
diff.copyTo(B(Rect(0, idx*3, 1, 3)));
```
  - `Matx` and `Mat` cannot be mixed. For instance with matrix multiplication:
```
    Mat M1 = (Mat_<double>(4,4,CV_64FC1) << 1, 2, 3, 4,
             5, 6, 7, 8,
             9, 10, 11, 12,
             13, 14, 15, 16);

    Matx44d M2(1, 2, 3, 4,
               5, 6, 7, 8,
               9, 10, 11, 12,
               13, 14, 15, 16);
    std::cout << "M1*M1:\n" << M1*M1 << std::endl; //OK
    std::cout << "M2*M2:\n" << M2*M2 << std::endl; //OK
    std::cout << "M1*M2:\n" << M1*M2 << std::endl; //NOK
    std::cout << "M1*Mat(M2):\n" << M1*Mat(M2) << std::endl; //OK
```

Also, I have found a weird thing, probably a missing operator:
```
cv::Matx33d skew2(const cv::Matx31d& v)
{
  double vx = v(0,0);
  double vy = v(1,0);
  double vz = v(2,0);
  return Matx33d(0, -vz, vy,
                 vz, 0, -vx,
                 -vy, vx, 0);
}

int main()
{
    Matx31d P1(1, 2, 3);
    std::cout << "skew(P1/2):\n" << skew2(P1/2) << std::endl; //NOK
    std::cout << "skew(P1*2):\n" << skew2(P1*2) << std::endl; //OK
}
```

---

Some functions that I had to defined but are worth or could be available in OpenCV:
  - skew matrix from a vector
  - conversion functions between rotation matrix and quaternion, quaternion type like in Eigen, quaternion multiplication, ...?
  - homogeneous matrix inverse?
  - kronecker product, I had to take the implementation from [here](https://stackoverflow.com/questions/36549057/opencv-tensor-product/36552682#36552682). Since it should be also the same thing than tensor product, maybe there is already this operation in the DNN module?

**Comment by alalek** [comment]
@catree Could you please fill an issue about Matx multiplication with your code? I will try to look on it later.

File: modules/calib3d/include/opencv2/calib3d.hpp [hunk 2]
```diff
@@ -1402,6 +1410,139 @@ CV_EXPORTS_W Mat getOptimalNewCameraMatrix( InputArray cameraMatrix, InputArray
...
+Additional information can be found on this [website](http://campar.in.tum.de/Chair/HandEyeCalibration).
+\note
+A minimum of 2 motions with non parallel rotation axes are necessary to determine the hand-eye transformation.
+So at least 3 different poses are required, but it is strongly recommended to use many more poses.
+
+ */
+CV_EXPORTS_W void calibrateHandEye( InputArrayOfArrays R_gripper2base, InputArrayOfArrays t_gripper2base,
+                                    InputArrayOfArrays R_target2cam, InputArrayOfArrays t_target2cam,
+                                    OutputArray R_cam2gripper, OutputArray t_cam2gripper,
+                                    int method=CALIB_HAND_EYE_TSAI );
```
Comment (alalek) [inline modules/calib3d/include/opencv2/calib3d.hpp]: Consider using enum HandEyeCalibrationMethod here (but need to check that bindings work properly).

File: modules/calib3d/src/calibration_handeye.cpp [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+
+    Mat Pcg_;
+    //Rotation from camera to gripper is obtained from the set of equations:
+    //    skew(Pgij+Pcij) * Pcg_ = Pcij - Pgij    (eq 12)
+    solve(A, B, Pcg_, DECOMP_SVD);
+
+    Mat Pcg_norm = Pcg_.t() * Pcg_;
+    //Obtained non-unit quaternion is scaled back to unit value that
+   //designates camera-gripper rotation
+    Mat Pcg = 2 * Pcg_ / sqrt(1 + Pcg_norm.at<double>(0,0)); //eq 14
```
Comment (alalek) [inline modules/calib3d/src/calibration_handeye.cpp]: indentation

File: modules/calib3d/src/calibration_handeye.cpp [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+//Matlab code: http://math.loyola.edu/~mili/Calibration/
+static void calibrateHandEyePark(const std::vector<Mat>& Hg, const std::vector<Mat>& Hc,
+                                 Mat& R_cam2gripper, Mat& t_cam2gripper)
+{
+    Mat M = Mat::zeros(3, 3, CV_64FC1);
+
+    int idx = 0;
+    for (size_t i = 0; i < Hg.size(); i++)
+    {
+        for (size_t j = i+1; j < Hg.size(); j++, idx++)
```
Comment (alalek) [inline modules/calib3d/src/calibration_handeye.cpp]: Looks like `idx` is not used here.

File: modules/calib3d/src/calibration_handeye.cpp [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+//Matlab code: http://math.loyola.edu/~mili/Calibration/
+static void calibrateHandEyeHoraud(const std::vector<Mat>& Hg, const std::vector<Mat>& Hc,
+                                   Mat& R_cam2gripper, Mat& t_cam2gripper)
+{
+    Mat A = Mat::zeros(4, 4, CV_64FC1);
+
+    int idx = 0;
+    for (size_t i = 0; i < Hg.size(); i++)
+    {
+        for (size_t j = i+1; j < Hg.size(); j++, idx++)
```
Comment (alalek) [inline modules/calib3d/src/calibration_handeye.cpp]: idx - used?

File: modules/calib3d/src/calibration_handeye.cpp [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+    }
+
+    Mat X;
+    solve(A, B, X, DECOMP_SVD);
+
+    Mat R = X(Rect(0, 0, 1, 9));
+    int newSize[] = {3, 3};
+    R = R.reshape(1, 2, newSize);
+    //Eq 15
+    R = pow(sign_double(determinant(R)) / abs(determinant(R)), 1.0/3.0) * R;
```
Comment (alalek) [inline modules/calib3d/src/calibration_handeye.cpp]: > determinant(R)

It is better to "cache" this value in local variable.

**Comment by catree** [comment]
Thanks for the review.

Build should fail when using enum. I got this:

```
In file included from /home/OpenCV-build/modules/java_bindings_generator/gen/cpp/opencv_jni.hpp:10:0,
                 from /home/OpenCV-src/modules/java/generator/src/cpp/opencv_java.cpp:8:
/home/OpenCV-build/modules/java_bindings_generator/gen/cpp/calib3d.inl.hpp: In function ‘void Java_org_opencv_calib3d_Calib3d_calibrateHandEye_12(JNIEnv*, jclass, jlong, jlong, jlong, jlong, jlong)’:
/home/OpenCV-build/modules/java_bindings_generator/gen/cpp/calib3d.inl.hpp:4662:105: error: too few arguments to function ‘void cv::calibrateHandEye(cv::InputArrayOfArrays, cv::InputArrayOfArrays, cv::InputArrayOfArrays, cv::InputArrayOfArrays, cv::OutputArray, cv::OutputArray, cv::HandEyeCalibrationMethod)’
         cv::calibrateHandEye( R_gripper2base, t_gripper2base, R_target2cam, t_target2cam, R_cam2gripper );
                                                                                                         ^
In file included from /home/OpenCV-build/modules/java_bindings_generator/gen/cpp/calib3d.inl.hpp:12:0,
                 from /home/OpenCV-build/modules/java_bindings_generator/gen/cpp/opencv_jni.hpp:10,
                 from /home/OpenCV-src/modules/java/generator/src/cpp/opencv_java.cpp:8:
/home/OpenCV-src/modules/calib3d/include/opencv2/calib3d.hpp:1541:47: note: declared here
 CV_EXPORTS_W void calibrateHandEye( InputArrayOfArrays R_gripper2base, InputArrayOfArrays t_gripper2base,
```

**Comment by msnh2012** [comment]
@catree Can I put the params which from "solvePnp->Rodrigues" into "calibrateHandEye" directly? And I get the "rx ry rz" euler angles from my robot, which kinds of method can help me to get the rotation matrix which exactly match "calibrateHandEye"?

**Comment by catree** [comment]
@msnh2012 

For `R_target2cam`, you can use `findChessboardCorners()` to get the chessboard corners and `solvePnP()` to estimate the chessboard pose with respect to the camera frame. `solvePnP()` returns a Rodrigues rotation vector so you can use `Rodrigues()` to get the corresponding rotation matrix.

For `R_gripper2base`, if your robot framework returns "rx, ry, rz" Euler angles, you have to check the manufacturer documentation to properly convert the Euler angles to the rotation matrix according to the chosen [convention](https://en.wikipedia.org/wiki/Euler_angles#Rotation_matrix).

The best would be to check if you can get directly either the rotation matrix or quaternion. With quaternion, the only pitfall is to be sure to get the correct "qx, qy,qz, qw" values.

**Comment by msnh2012** [comment]
@catree 感谢如此快的回复，从solvePnp到Rodrigues输出的旋转矩阵转成欧拉角，采用哪种旋向？Z-Y-X还是X-Y-Z？然后calibrateHandEye采用的是统一的Z-Y-X还是X-Y-Z旋向？还是只考虑旋转矩阵，没考虑由欧拉角转换得到的旋转矩阵这种情况？然后就是对于标定的结果，有没有一种误差的评定方法？谢谢:D

**Comment by catree** [comment]
@msnh2012 Please write in English.

`calibrateHandEye` needs `3x3` rotation matrix. Depending on what you have, the possible conversions are:

- Axis-angle to rotation matrix, done in `cv::Rodrigues()`
- Quaternion to rotation matrix
- Euler angles to rotation matrix

This is the representation of the Axis-Angle rotation formalism, also called rotation vector and also called Rodrigues vector:

image

You should read a texbook on this subject, for [instance](https://www.eecis.udel.edu/~cer/arv/readings/old_mkss.pdf). There are plenty of useful ressources on the net if you want more information.

**Comment by msnh2012** [comment]
@catree Thanks!  Can u give an example of hand-eye calibration for us? That will be very nice .

**Comment by catree** [comment]
@msnh2012 

You can find [here](https://github.com/opencv/opencv/blob/4.1.0/modules/calib3d/test/test_calibration_hand_eye.cpp) some example code.

**Comment by msnh2012** [comment]
@catree  Thanks! :+1:

**Comment by zarathustr** [comment]
> ### This pullrequest changes
> Add Hand-Eye calibration from Tsai method. This allows estimating the rigid transformation between a camera mounted on a robot gripper:
> 
> hand-eye_figure
> 
> See also:
> 
> * [Hand-Eye Calibration ](http://campar.in.tum.de/Chair/HandEyeCalibration)
> * [A new technique for fully autonomous and efficient 3D robotics hand/eye calibration,  R.Y. Tsai ; R.K. Lenz](https://pdfs.semanticscholar.org/19b3/89a797a55c8b63dca8b6d1889df4cff8bfaa.pdf)
> 
> The current interface is:
> 
> ```
> void calibrateHandEyeTsai( InputArrayOfArrays R_gripper2base, InputArrayOfArrays t_gripper2base,
>                                             InputArrayOfArrays R_target2cam, InputArrayOfArrays t_target2cam,
>                                             OutputArray R_cam2gripper, OutputArray t_cam2gripper );
> ```
> 
> * currently there is only one method (Tsai). Should we keep this or use a more generic one `calibrateHandEye` with a method parameter to choose different calibration methods? This can be a more general question also (e.g. `calibrateCamera`, `calibrateCameraRO`, `fitEllipse`, `fitEllipseAMS`, `fitEllipseDirect`, ...).
> * inputs are rotation matrices and translation vectors. It could have been rotation and translation vectors
> * naming is consistent with the other functions, alternative is `HandEyeCalibration()`
> * code follows the paper notations
> * C++ code from this [Matlab code](http://lazax.com/www.cs.columbia.edu/~laza/html/Stewart/matlab/handEye.m)

Is it necessary to include more hand-eye calibration methods? It seems that most of these analytical methods do not work well for real-world tasks.

**Comment by catree** [comment]
@zarathustr 
I think that the best would be to have a non-linear optimization method. The few methods I had seen were using the Ceres solver. But I have no experience with it, and it can be tricky to integrate in OpenCV.

Feel free to contribute in OpenCV (BSD-3 license) if you have better methods.

**Pull request merged**